crates/wasm-encoder/src/component/types.rs

```diff
@@ -0,0 +1,574 @@
+                ty.encode(bytes)
+            }
+            Self::Alias(alias) => {
+                bytes.push(MODULE_TYPEDEF_ALIAS);
+                alias.encode(bytes);
+            }
+            Self::Export(name, ty) => {
+                bytes.push(MODULE_TYPEDEF_EXPORT);
+                bytes.extend(encoders::str(name));
+                ty.encode(bytes);
+            }
+            Self::Import(name, ty) => {
+                bytes.push(MODULE_TYPEDEF_IMPORT);
+                bytes.extend(encoders::str(name));
+                ty.encode(bytes);
+            }
+        }
+    }
+}
+
+/// Represents an interface type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum InterfaceType {
+    /// The type is a boolean.
+    Bool,
+    /// The type is a signed 8-bit integer.
+    S8,
+    /// The type is an unsigned 8-bit integer.
+    U8,
+    /// The type is a signed 16-bit integer.
+    S16,
+    /// The type is an unsigned 16-bit integer.
+    U16,
+    /// The type is a signed 32-bit integer.
+    S32,
+    /// The type is an unsigned 32-bit integer.
+    U32,
+    /// The type is a signed 64-bit integer.
+    S64,
+    /// The type is an unsigned 64-bit integer.
+    U64,
+    /// The type is a 32-bit floating point number.
+    F32,
+    /// The type is a 64-bit floating point number.
+    F64,
+    /// The type is a Unicode character.
+    Char,
+    /// The type is a string.
+    String,
+    /// The type is the unit type.
+    Unit,
+    /// The type is a compound interface type.
+    ///
+    /// The value is a type index to a compound type.
+    Compound(u32),
+}
+
+impl InterfaceType {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::Bool => bytes.push(INTERFACE_TYPE_BOOL),
+            Self::S8 => bytes.push(INTERFACE_TYPE_S8),
+            Self::U8 => bytes.push(INTERFACE_TYPE_U8),
+            Self::S16 => bytes.push(INTERFACE_TYPE_S16),
+            Self::U16 => bytes.push(INTERFACE_TYPE_U16),
+            Self::S32 => bytes.push(INTERFACE_TYPE_S32),
+            Self::U32 => bytes.push(INTERFACE_TYPE_U32),
+            Self::S64 => bytes.push(INTERFACE_TYPE_S64),
+            Self::U64 => bytes.push(INTERFACE_TYPE_U64),
+            Self::F32 => bytes.push(INTERFACE_TYPE_F32),
+            Self::F64 => bytes.push(INTERFACE_TYPE_F64),
+            Self::Char => bytes.push(INTERFACE_TYPE_CHAR),
+            Self::String => bytes.push(INTERFACE_TYPE_STRING),
+            Self::Unit => bytes.push(INTERFACE_TYPE_UNIT),
+            Self::Compound(index) => bytes.extend(encoders::u32(*index)),
+        }
+    }
+}
+
+/// Represents a compound interface type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CompoundInterfaceType<'a> {
+    /// A list interface type.
+    List(InterfaceType),
+    /// A record interface type.
+    Record(&'a [(&'a str, InterfaceType)]),
+    /// A variant interface type.
+    Variant(&'a [(&'a str, InterfaceType)]),
```

Comment:
I think that the interface type here is optional?

Reply:
Good point. I think that needs a change to the binary format to support, as we should be able to support both case `Foo` and case `Foo(())` right?

Reply:
Actually, I don't think we can actually represent `Foo(())` in `wit` currently, right? Still, can't hurt to change the format to make the interface type optional.

Reply:
We've updated the spec to make the types in variant optional, plus unified the expected representation so that both `ok` and `error` cases have optional types. The unit type was removed from the spec for now.